File: rug-pull-detector/python/rug-pull-detector.py
from web3 import Web3
import datetime
import io
import sys
from decimal import Decimal

uniswap_v2_factory_abi = [  # Minimal ABI for Factory contract
    {
        "constant": True,
        "inputs": [{"name": "tokenA", "type": "address"}, {"name": "tokenB", "type": "address"}],
        "name": "getPair",
        "outputs": [{"name": "", "type": "address"}],
        "payable": False,
        "stateMutability": "view",
        "type": "function"
    },
    {
        "constant": True,
        "inputs": [],
        "name": "allPairsLength",
        "outputs": [{"name": "", "type": "uint256"}],
        "payable": False,
        "stateMutability": "view",
        "type": "function"
    },
    {
        "constant": True,
        "inputs": [{"name": "uint", "type": "uint256"}],
        "name": "allPairs",
        "outputs": [{"name": "pair", "type": "address"}],
        "payable": False,
        "stateMutability": "view",
        "type": "function"
    }
]

pair_abi = [
    {"constant": True, "inputs": [], "name": "token0", "outputs": [{"name": "", "type": "address"}],
     "stateMutability": "view", "type": "function"},
    {"constant": True, "inputs": [], "name": "token1", "outputs": [{"name": "", "type": "address"}],
     "stateMutability": "view", "type": "function"},
    {"constant": True, "inputs": [], "name": "totalSupply", "outputs": [{"name": "", "type": "uint256"}],
     "stateMutability": "view", "type": "function"},
    {"constant": True, "inputs": [], "name": "getReserves",
     "outputs": [{"name": "reserve0", "type": "uint112"}, {"name": "reserve1", "type": "uint112"},
                 {"name": "blockTimestampLast", "type": "uint32"}], "stateMutability": "view", "type": "function"},
    {"constant": True, "inputs": [], "name": "decimals", "outputs": [{"name": "", "type": "uint8"}],
     "stateMutability": "view", "type": "function"}
]

token_abi = [
    {"constant": True, "inputs": [], "name": "name", "outputs": [{"name": "", "type": "string"}],
     "stateMutability": "view", "type": "function"},
    {"constant": True, "inputs": [], "name": "symbol", "outputs": [{"name": "", "type": "string"}],
     "stateMutability": "view", "type": "function"},
    {
        "constant": True,
        "inputs": [],
        "name": "totalSupply",
        "outputs": [{"name": "", "type": "uint256"}],
        "stateMutability": "view",
        "type": "function"
    },
    {"constant": True, "inputs": [], "name": "decimals", "outputs": [{"name": "", "type": "uint8"}],
     "stateMutability": "view", "type": "function"}
]

BASE_RPC_URL = "https://base.drpc.org"  # Base RPC URL
web3 = Web3(Web3.HTTPProvider(BASE_RPC_URL))

# Ensure the connection to the Base chain
if web3.is_connected():
    print("Connected to Base Chain")

# Input token address 
input_token_address = "0x768BE13e1680b5ebE0024C42c896E3dB59ec0149"

# Uniswap V2 Factory contract address 
uniswap_v2_factory_address = "0x8909Dc15e40173Ff4699343b6eB8132c65e18eC6"

# Do not change these. Will be used to find the pair with USDC or WETH
USDC_contract = '0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913'
WETH_contract = '0x4200000000000000000000000000000000000006'
ZERO_address = '0x0000000000000000000000000000000000000000'

# Create contract instance for the Uniswap V2 Factory
factory_contract = web3.eth.contract(address=web3.to_checksum_address(uniswap_v2_factory_address),
                                     abi=uniswap_v2_factory_abi)

def get_token_decimals(token_address):
    if token_address == USDC_contract:
        return 6
    if token_address == WETH_contract:
        return 18
    
    token_contract = web3.eth.contract(address=web3.to_checksum_address(token_address), abi=token_abi)
    return token_contract.functions.decimals().call()

def get_token_total_supply(token_address):
    token_contract = web3.eth.contract(address=web3.to_checksum_address(token_address), abi=token_abi)
    return token_contract.functions.totalSupply().call()

def check_minting_ability(token_contract):
    try:
        token_contract.functions.mint().call()
        return {"mintable": True, "supplyStatus": "NOT FIXED"}
    except:
        return {"mintable": False, "supplyStatus": "FIXED"}

def find_pair_by_token(token_address):
    pair_address_usdc = factory_contract.functions.getPair(
        web3.to_checksum_address(token_address),
        web3.to_checksum_address(USDC_contract)
    ).call()
    
    if pair_address_usdc != ZERO_address:
        return {"pairAddress": pair_address_usdc, "quoteToken": "USDC"}
    
    pair_address_weth = factory_contract.functions.getPair(
        web3.to_checksum_address(token_address),
        web3.to_checksum_address(WETH_contract)
    ).call()
    
    if pair_address_weth != ZERO_address:
        return {"pairAddress": pair_address_weth, "quoteToken": "WETH"}
    
    return None

def calculate_market_cap(pair_contract):
    try:
        reserves = pair_contract.functions.getReserves().call()
        token0 = pair_contract.functions.token0().call()
        token1 = pair_contract.functions.token1().call()
        
        token_total_supply = get_token_total_supply(token0)
        
        # Normalize values using decimals
        reserve0_normalized = Decimal(reserves[0]) / Decimal(10 ** get_token_decimals(token0))
        reserve1_normalized = Decimal(reserves[1]) / Decimal(10 ** get_token_decimals(token1))
        price_per_token = reserve1_normalized / reserve0_normalized
        total_supply_normalized = Decimal(token_total_supply) / Decimal(10 ** get_token_decimals(token0))
        market_cap = total_supply_normalized * price_per_token
        
        return {
            "reserves": reserves,
            "token0": token0,
            "token1": token1,
            "pricePerToken": str(price_per_token),
            "totalSupplyNormalized": str(total_supply_normalized),
            "marketCap": str(market_cap)
        }
    except Exception as e:
        print(f"Error calculating market cap: {e}")
        return None

# Main execution
pair_info = find_pair_by_token(input_token_address)
if pair_info is None:
    print("Pair could not be found! Quit execution of the algorithm...")
    quit()

pair_address = pair_info["pairAddress"]
pair_contract = web3.eth.contract(address=web3.to_checksum_address(pair_address), abi=pair_abi)

# Get token addresses
token0 = pair_contract.functions.token0().call()
token1 = pair_contract.functions.token1().call()
input_token = token0 if token0 != USDC_contract and token0 != WETH_contract else token1
pair_token = token0 if token0 == USDC_contract or token0 == WETH_contract else token1

# Create contract instances
token_contract = web3.eth.contract(address=input_token, abi=token_abi)
pair_token_contract = web3.eth.contract(address=pair_token, abi=token_abi)

buffer = io.StringIO()
sys.stdout = buffer

print("=" * 80)
print("🔍 TOKEN ANALYSIS REPORT")
print("=" * 80)
print(f"Generated on: {datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
print("-" * 80)

# Token Information
print("\n📊 TOKEN INFORMATION")
print("-" * 40)
print(f"Token Address: {input_token}")
token_name = token_contract.functions.name().call()
token_symbol = token_contract.functions.symbol().call()
print(f"Token Name: {token_name}")
print(f"Token Symbol: {token_symbol}")

# Pair Token Information
print("\n🔄 PAIR TOKEN INFORMATION")
print("-" * 40)
print(f"Pair Token Address: {pair_token}")
pair_token_name = pair_token_contract.functions.name().call()
pair_token_symbol = pair_token_contract.functions.symbol().call()
print(f"Pair Token Name: {pair_token_name}")
print(f"Pair Token Symbol: {pair_token_symbol}")

# Liquidity Information
print("\n💧 LIQUIDITY INFORMATION")
print("-" * 40)
print(f"Liquidity Pair Address: {pair_address}")
total_lp_tokens = pair_contract.functions.totalSupply().call()
print(f"Total Supply of LP Tokens: {total_lp_tokens}")

# Market Analysis
print("\n💰 MARKET ANALYSIS")
print("-" * 40)
market_cap_data = calculate_market_cap(pair_contract)
if market_cap_data:
    print(f"Reserve {token_symbol}: {market_cap_data['reserves'][0]}")
    print(f"Reserve {pair_token_symbol}: {market_cap_data['reserves'][1]}")
    print(f"Price per {token_symbol}: {market_cap_data['pricePerToken']} {pair_token_symbol}")
    print(f"Total Supply: {market_cap_data['totalSupplyNormalized']}")
    print(f"Market Cap: {market_cap_data['marketCap']} {pair_token_symbol}")

# Supply Analysis
print("\n🪄 SUPPLY ANALYSIS")
print("-" * 40)
minting_status = check_minting_ability(token_contract)
print(f"Mint Status: {'MINTABLE' if minting_status['mintable'] else 'NOT MINTABLE'}")
print(f"Total Supply Status: {minting_status['supplyStatus']}")

print("\n" + "=" * 80)
print("End of Report")
print("=" * 80)

sys.stdout = sys.__stdout__

output_text = buffer.getvalue()

# Save to text file
txt_filename = '/data/outputs/report.txt'
with open(txt_filename, 'w') as f:
    f.write(output_text)

print(f"✅ Output saved to {txt_filename} \n")
print(output_text)

File: rug-pull-detector/node/rug-pull-detector.js
const { Web3 } = require("web3");
const fs = require("fs");
const path = require("path");
const Bignumber = require("bignumber.js");

// ============= CONFIGURATION =============
const CONFIG = {
  RPC_URL: "https://base.drpc.org",
  CONTRACTS: {
    USDC: "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
    WETH: "0x4200000000000000000000000000000000000006",
    ZERO: "0x0000000000000000000000000000000000000000",
    UNISWAP_V2_FACTORY: "0x8909Dc15e40173Ff4699343b6eB8132c65e18eC6",
  },
  OUTPUT_DIR: "/data/outputs",
};

// ============= CONTRACT ABIs =============
const ABIs = {
  uniswapV2Factory: [
    {
      constant: true,
      inputs: [
        { name: "tokenA", type: "address" },
        { name: "tokenB", type: "address" },
      ],
      name: "getPair",
      outputs: [{ name: "", type: "address" }],
      payable: false,
      stateMutability: "view",
      type: "function",
    },
    {
      constant: true,
      inputs: [],
      name: "allPairsLength",
      outputs: [{ name: "", type: "uint256" }],
      payable: false,
      stateMutability: "view",
      type: "function",
    },
    {
      constant: true,
      inputs: [{ name: "uint", type: "uint256" }],
      name: "allPairs",
      outputs: [{ name: "pair", type: "address" }],
      payable: false,
      stateMutability: "view",
      type: "function",
    },
  ],
  pair: [
    {
      constant: true,
      inputs: [],
      name: "token0",
      outputs: [{ name: "", type: "address" }],
      stateMutability: "view",
      type: "function",
    },
    {
      constant: true,
      inputs: [],
      name: "token1",
      outputs: [{ name: "", type: "address" }],
      stateMutability: "view",
      type: "function",
    },
    {
      constant: true,
      inputs: [],
      name: "totalSupply",
      outputs: [{ name: "", type: "uint256" }],
      stateMutability: "view",
      type: "function",
    },
    {
      constant: true,
      inputs: [],
      name: "getReserves",
      outputs: [
        { name: "reserve0", type: "uint112" },
        { name: "reserve1", type: "uint112" },
        { name: "blockTimestampLast", type: "uint32" },
      ],
      stateMutability: "view",
      type: "function",
    },
    {
      constant: true,
      inputs: [],
      name: "decimals",
      outputs: [{ name: "", type: "uint8" }],
      stateMutability: "view",
      type: "function",
    },
  ],
  token: [
    {
      constant: true,
      inputs: [],
      name: "name",
      outputs: [{ name: "", type: "string" }],
      stateMutability: "view",
      type: "function",
    },
    {
      constant: true,
      inputs: [],
      name: "symbol",
      outputs: [{ name: "", type: "string" }],
      stateMutability: "view",
      type: "function",
    },
    {
      constant: true,
      inputs: [],
      name: "totalSupply",
      outputs: [{ name: "", type: "uint256" }],
      stateMutability: "view",
      type: "function",
    },
    {
      constant: true,
      inputs: [],
      name: "decimals",
      outputs: [{ name: "", type: "uint8" }],
      stateMutability: "view",
      type: "function",
    },
  ],
};

// ============= WEB3 SETUP =============
const web3 = new Web3(CONFIG.RPC_URL);
const factoryContract = new web3.eth.Contract(
  ABIs.uniswapV2Factory,
  CONFIG.CONTRACTS.UNISWAP_V2_FACTORY
);

// ============= UTILITY FUNCTIONS =============
class TokenUtils {
  static async getTokenDecimals(tokenAddress) {
    if (tokenAddress === CONFIG.CONTRACTS.USDC) return 6;
    if (tokenAddress === CONFIG.CONTRACTS.WETH) return 18;

    const tokenContract = new web3.eth.Contract(ABIs.token, tokenAddress);
    return await tokenContract.methods.decimals().call();
  }

  static async getTokenTotalSupply(tokenAddress) {
    const tokenContract = new web3.eth.Contract(ABIs.token, tokenAddress);
    return await tokenContract.methods.totalSupply().call();
  }

  static async checkMintingAbility(tokenContract) {
    try {
      await tokenContract.methods.mint().call();
      return { mintable: true, supplyStatus: "NOT FIXED" };
    } catch {
      return { mintable: false, supplyStatus: "FIXED" };
    }
  }
}

// ============= CORE FUNCTIONS =============
class TokenAnalyzer {
  static async findPairByToken(tokenAddress) {
    try {
      const pairAddressUsdc = await factoryContract.methods
        .getPair(
          web3.utils.toChecksumAddress(tokenAddress),
          web3.utils.toChecksumAddress(CONFIG.CONTRACTS.USDC)
        )
        .call();

      if (pairAddressUsdc !== CONFIG.CONTRACTS.ZERO) {
        return { pairAddress: pairAddressUsdc, quoteToken: "USDC" };
      }

      const pairAddressWeth = await factoryContract.methods
        .getPair(
          web3.utils.toChecksumAddress(tokenAddress),
          web3.utils.toChecksumAddress(CONFIG.CONTRACTS.WETH)
        )
        .call();

      if (pairAddressWeth !== CONFIG.CONTRACTS.ZERO) {
        return { pairAddress: pairAddressWeth, quoteToken: "WETH" };
      }

      return null;
    } catch (error) {
      console.error("Error finding pair:", error);
      return null;
    }
  }

  static async calculateMarketCap(pairContract) {
    try {
      const [reserves, token0, token1] = await Promise.all([
        pairContract.methods.getReserves().call(),
        pairContract.methods.token0().call(),
        pairContract.methods.token1().call(),
      ]);

      const tokenTotalSupply = await TokenUtils.getTokenTotalSupply(token0);

      const reserve0Normalized = new Bignumber(reserves[0]).div(
        new Bignumber(10).pow(await TokenUtils.getTokenDecimals(token0))
      );
      const reserve1Normalized = new Bignumber(reserves[1]).div(
        new Bignumber(10).pow(await TokenUtils.getTokenDecimals(token1))
      );
      const pricePerToken = reserve1Normalized.div(reserve0Normalized);
      const totalSupplyNormalized = new Bignumber(tokenTotalSupply).div(
        new Bignumber(10).pow(await TokenUtils.getTokenDecimals(token0))
      );
      const marketCap = totalSupplyNormalized.multipliedBy(pricePerToken);

      return {
        reserves,
        token0,
        token1,
        pricePerToken: pricePerToken.toString(),
        totalSupplyNormalized: totalSupplyNormalized.toString(),
        marketCap: marketCap.toString(),
      };
    } catch (error) {
      console.log(`Error calculating market cap: ${error.message}`);
      return null;
    }
  }
}

// ============= REPORT GENERATION =============
class ReportGenerator {
  static async generateReport(inputTokenAddress) {
    const output = [];

    // Header
    output.push("=".repeat(80));
    output.push("🔍 TOKEN ANALYSIS REPORT");
    output.push("=".repeat(80));
    output.push(`Generated on: ${new Date().toLocaleString()}`);
    output.push("-".repeat(80));

    const pairInfo = await TokenAnalyzer.findPairByToken(inputTokenAddress);
    if (!pairInfo) {
      output.push(
        "Pair could not be found! Quit execution of the algorithm..."
      );
      return;
    }

    const pairAddress = pairInfo.pairAddress;
    const pairContract = new web3.eth.Contract(ABIs.pair, pairAddress);
    const token0 = await pairContract.methods.token0().call();
    const token1 = await pairContract.methods.token1().call();
    const inputToken =
      token0 !== CONFIG.CONTRACTS.USDC && token0 !== CONFIG.CONTRACTS.WETH
        ? token0
        : token1;
    const pairToken =
      token0 === CONFIG.CONTRACTS.USDC || token0 === CONFIG.CONTRACTS.WETH
        ? token0
        : token1;

    const tokenContract = new web3.eth.Contract(ABIs.token, inputToken);
    const pairTokenContract = new web3.eth.Contract(ABIs.token, pairToken);

    // Token Information
    output.push("\n📊 TOKEN INFORMATION");
    output.push("-".repeat(40));
    output.push(`Token Address: ${inputToken}`);
    const tokenName = await tokenContract.methods.name().call();
    const tokenSymbol = await tokenContract.methods.symbol().call();
    output.push(`Token Name: ${tokenName}`);
    output.push(`Token Symbol: ${tokenSymbol}`);

    // Pair Token Information
    output.push("\n🔄 PAIR TOKEN INFORMATION");
    output.push("-".repeat(40));
    output.push(`Pair Token Address: ${pairToken}`);
    const pairTokenName = await pairTokenContract.methods.name().call();
    const pairTokenSymbol = await pairTokenContract.methods.symbol().call();
    output.push(`Pair Token Name: ${pairTokenName}`);
    output.push(`Pair Token Symbol: ${pairTokenSymbol}`);

    // Liquidity Information
    output.push("\n💧 LIQUIDITY INFORMATION");
    output.push("-".repeat(40));
    output.push(`Liquidity Pair Address: ${pairAddress}`);
    const totalLpTokens = await pairContract.methods.totalSupply().call();
    output.push(`Total Supply of LP Tokens: ${totalLpTokens}`);

    // Market Analysis
    output.push("\n💰 MARKET ANALYSIS");
    output.push("-".repeat(40));
    const marketCapData = await TokenAnalyzer.calculateMarketCap(pairContract);
    if (marketCapData) {
      output.push(`Reserve ${tokenSymbol}: ${marketCapData.reserves[0]}`);
      output.push(`Reserve ${pairTokenSymbol}: ${marketCapData.reserves[1]}`);
      output.push(
        `Price per ${tokenSymbol}: ${marketCapData.pricePerToken} ${pairTokenSymbol}`
      );
      output.push(`Total Supply : ${marketCapData.totalSupplyNormalized}`);
      output.push(`Market Cap: ${marketCapData.marketCap} ${pairTokenSymbol}`);
    }

    // Supply Analysis
    output.push("\n🪄 SUPPLY ANALYSIS");
    output.push("-".repeat(40));

    const mintingStatus = await TokenUtils.checkMintingAbility(tokenContract);
    output.push(
      `Mint Status: ${mintingStatus.mintable ? "MINTABLE" : "NOT MINTABLE"}`
    );
    output.push(`Total Supply Status: ${mintingStatus.supplyStatus}`);

    output.push("\n" + "=".repeat(80));
    output.push("End of Report");
    output.push("=".repeat(80));

    // Save to file
    const outputText = output.join("\n");
    if (!fs.existsSync(CONFIG.OUTPUT_DIR)) {
      fs.mkdirSync(CONFIG.OUTPUT_DIR, { recursive: true });
    }

    const txtFilename = path.join(CONFIG.OUTPUT_DIR, "report.txt");
    fs.writeFileSync(txtFilename, outputText);

    // Display the report in terminal
    console.log(outputText);
    console.log(`✅ Output saved to ${txtFilename}\n`);
  }
}

// ============= MAIN EXECUTION =============
const inputTokenAddress = "0x768BE13e1680b5ebE0024C42c896E3dB59ec0149";
ReportGenerator.generateReport(inputTokenAddress).catch(console.error);
